src/Npgsql/NpgsqlConnector.cs

```diff
@@ -860,8 +860,12 @@ await socket.ConnectAsync(endpoint)
 
                     if (perIpTimeout.IsSet)
                     {
+                        var connectTimeout = (int)perIpTimeout.TimeLeft.TotalMilliseconds;
```

Comment:
This seems like it just moved the problem downline. if an exception is force thrown of the token throws one same result. would an approach be to not set the CancelAfter if its < 0?

Reply:
That's kind of its point - we have a timer to open the connection. If we go over it, we should get an exception. In your case, I would suggest raising the connection string's `Timeout` property from default 15 seconds up to 30 (or maybe even more).